smithy-typescript-codegen/src/main/java/software/amazon/smithy/typescript/codegen/ServerCommandGenerator.java

```diff
@@ -99,4 +113,95 @@ private void writeOutputType(String typeName, Optional<StructureShape> outputSha
+        writer.write("export type $L = __Operation<$T, $T>", operationSymbol.getName(), inputType, outputType);
+        writer.write("");
+    }
+
+    private void writeOperationSerializer() {
+        Symbol operationSymbol = symbolProvider.toSymbol(operation);
+        String serializerName = operationSymbol.expectProperty("serializerType", Symbol.class).getName();
+        Symbol serverSymbol = symbolProvider.toSymbol(model.expectShape(settings.getService()));
+
+        writer.addImport("OperationSerializer", null, "@aws-smithy/server-common");
+        writer.openBlock("export class $L implements OperationSerializer<$T, $S, $T> {", "}",
+                serializerName, serverSymbol, operation.getId().getName(), errorsType, () -> {
+            String serializerFunction = ProtocolGenerator.getGenericSerFunctionName(operationSymbol) + "Response";
+            String deserializerFunction = ProtocolGenerator.getGenericDeserFunctionName(operationSymbol) + "Request";
+            writer.addImport(serializerFunction, null,
+                    "./protocols/" + ProtocolGenerator.getSanitizedName(protocolGenerator.getName()));
+            writer.addImport(deserializerFunction, null,
+                    "./protocols/" + ProtocolGenerator.getSanitizedName(protocolGenerator.getName()));
+            writer.write("serialize = $L;", serializerFunction);
+            writer.write("deserialize = $L;", deserializerFunction);
+            writer.write("");
+            writeErrorChecker();
+            writeErrorHandler();
+        });
+        writer.write("");
+    }
+
+    private void writeErrorChecker() {
+        writer.openBlock("isOperationError(error: any): error is $T {", "};", errorsType, () -> {
+            if (operation.getErrors().isEmpty()) {
+                writer.write("return false;");
+            } else {
+                writer.writeInline("const names: $T['name'][] = [", errorsType);
+                for (Iterator<ShapeId> iter = operation.getErrors().iterator(); iter.hasNext();) {
+                    writer.writeInline("$S", iter.next().getName());
+                    if (iter.hasNext()) {
+                        writer.writeInline(", ");
+                    }
+                }
+                writer.write("];");
+                writer.write("return names.includes(error.name);");
+            }
+        });
+        writer.write("");
+    }
+
+    private void writeErrorHandler() {
+        writer.addImport("SerdeContext", null, "@aws-sdk/types");
+        writer.openBlock("serializeError(error: $T, ctx: Omit<SerdeContext, 'endpoint'>): Promise<$T> {", "}",
+                errorsType, applicationProtocol.getResponseType(), () -> {
+            writer.openBlock("switch (error.name) {", "}", () -> {
+                for (ShapeId errorId : operation.getErrors()) {
+                    writeErrorHandlerCase(errorId);
+                }
+                writer.openBlock("default: {", "}", () -> writer.write("throw error;"));
```

Comment:
In the end we will probably not want the framework to throw; instead it will render any error as a serialized InternalFailure (synthetic exception). So either this should default to serializing any error as an InternalFailure, or it should throw and we should have a higher-level catch all and a protocol-level utility function that serializes it.

Reply:
Yeah we will need some synthetic error handling. I omitted that for now to keep the pr focused on modeled errors. In the short term we can rely on APIG to catch any unhandled errors and return a 500.